src/main/java/com/google/api/codegen/configgen/DiscoInterfaceTransformer.java

```diff
@@ -0,0 +1,139 @@
+  }
+
+  @Override
+  public void generateResourceNameGenerations(ConfigNode parentNode, ApiModel model) {
+    Document document = ((DiscoApiModel) model).getDocument();
+    SetMultimap<String, String> resourceToNamePatternMap = getResourceToNamePatternMap(document);
+    FieldConfigNode resourceNameGenerationsNode =
+        MissingFieldTransformer.append("resource_name_generation", parentNode).generate();
+    if (NodeFinder.hasContent(resourceNameGenerationsNode.getChild())) {
+      return;
+    }
+
+    ConfigNode elementNode = new NullConfigNode();
+    ConfigNode prevNode = null;
+    for (Map.Entry<String, List<Method>> resource : document.resources().entrySet()) {
+      for (Method method : resource.getValue()) {
+        if (Strings.isNullOrEmpty(method.path())) {
+          continue;
+        }
+
+        int startLine = NodeFinder.getNextLine(prevNode == null ? parentNode : prevNode);
+        ConfigNode node = new ListItemConfigNode(startLine);
+        String messageName = DiscoGapicNamer.getRequestName(method).toUpperCamel();
+        ConfigNode messageNameNode =
+            FieldConfigNode.createStringPair(startLine, "message_name", messageName);
+        String parameterName =
+            DiscoGapicNamer.getResourceIdentifier(method.flatPath()).toLowerCamel();
+        String resourceName =
+            getResourceIdentifier(method, resource.getKey(), resourceToNamePatternMap);
+        ConfigNode fieldEntityMapNode =
+            new FieldConfigNode(NodeFinder.getNextLine(messageNameNode), "field_entity_map");
+        ConfigNode fieldEntityEntryNode =
+            FieldConfigNode.createStringPair(
+                NodeFinder.getNextLine(fieldEntityMapNode), parameterName, resourceName);
+        fieldEntityMapNode.setChild(fieldEntityEntryNode);
+        node.setChild(messageNameNode.insertNext(fieldEntityMapNode));
+
+        if (prevNode == null) {
+          resourceNameGenerationsNode.setChild(node);
+        } else {
+          prevNode.insertNext(node);
+        }
+
+        if (!elementNode.isPresent()) {
+          elementNode = node;
+        }
+
+        prevNode = node;
+      }
+    }
+  }
+
+  /** Map of base resource identifiers to all canonical name patterns that use that identifier. */
+  private SetMultimap<String, String> getResourceToNamePatternMap(Document document) {
+    ImmutableSetMultimap.Builder<String, String> resourceToNamePatternMapBuilder =
+        ImmutableSetMultimap.builder();
+    for (Method method : document.methods()) {
+      String namePattern = DiscoGapicNamer.getCanonicalPath(method);
+      String simpleResourceName =
+          DiscoGapicNamer.getResourceIdentifier(method.flatPath()).toLowerCamel();
+      resourceToNamePatternMapBuilder.put(simpleResourceName, namePattern);
+    }
+
+    return resourceToNamePatternMapBuilder.build();
+  }
+
+  /**
+   * Get the resource name for a method. Qualifies the resource name if it clashes with another
```

Comment:
Is "resource name" the same thing as what's called a "collection name" elsewhere in the toolchain? If so, let's call it "collection" for consistency.